Openhardwaremonitor

wimhaanstra opened the pull request:
## Description:
Adds a Open Hardware Monitor platform, that enables Home Assistant to connect to another host.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2824

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: openhardwaremonitor
    host: HOSTNAME_OF_OPENHARDWAREMONITOR
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Commit:
Open Hardware Monitor sensor

Platform which is able to connect to the JSON API of Open Hardware Monitor and adds sensors for the devices.

Commit:
Remove copyright in header, not needed.

homeassistant commented:
Hi @depl0y, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#8056).

Once you do that we will be able to review and accept this pull request.

Thanks!

aronsky commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,212 @@
+        timer.start()
+
+    def initialize(self):
+        """Initial parsing of the sensors and adding of deviced."""
```
*devices*

aronsky commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,212 @@
+        host = self._config.get(CONF_HOST)
+        port = self._config.get(CONF_PORT)
+        data_url = "http://%s:%d/data.json" % (host, port)
+        _LOGGER.info("Download from %s", data_url)
```
This will unnecessarily spam the log, not needed.

aronsky commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,212 @@
+        computer_count = len(self._data[OHM_CHILDREN])
+        _LOGGER.info("Detected %d PC(s)", computer_count)
+
+        if computer_count > 0:
```
Unnecessary, the `for` loop won't execute for `range(0, 0)` anyway.

aronsky commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,212 @@
+import logging
+import json
+
+import urllib.request
```
Consider using the `requests` module, it'll make reading remote JSON data a bit more straightforward.

aronsky commented:
Consider converting the code to asynchronous (using `asyncio` and `aiohttp`). Since it's a new platform, it's better to start off asynchronous, instead of converting it later.

On another note - if there is no Python library for accessing Open Hardware Monitor data, consider making one, publishing it, and using it in your code here.

depl0y commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,212 @@
+class OpenHardwareMonitorDevice(Entity):
+    """Device used to display information from OpenHardwareMonitor."""
+
+    def __init__(self, data, name, path, obj, attributes):
```
Problem is, this information is parsed on startup, but on an update this information isn't available anymore. I then use the supplied `path` argument to quickly parse the current/new values. I'll see what I can do to remove the parameters.

Commit:
- Removed old code
- Fixed typo’s in comments
- Removed log spamming
- Removed code that was unnecessary
- Use requests instead of urllib
- Moved sensor update functionality to data handler, to remove unwanted constructor parameters

Commit:
Fixed typo in comment
Added tests

Commit:
Added default fixture, to stabilize tests

Commit:
- Fix for values deeper than 4 levels, no longer relies on fixed level
- Fixed tests

pvizeli commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,192 @@
+"""Support for Open Hardware Monitor Sensor Platform."""
+
+from threading import Timer
```
Please use our internal core timer function from `helper/events.py`

depl0y replied:
Fixed in the latest commit. Thanks!

Commit:
Removed timer in preference of helper methods

pvizeli commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,195 @@
+        except requests.exceptions.ConnectionError:
+            _LOGGER.error("ConnectionError: Is OpenHardwareMonitor running?")
+
+    def schedule_retry(self):
```
Remove the retry logic. That make no sense. You set SCAN_INTERVAL to 3 sec to poll data. That is very hard. So you had allready 10 times reload before retry is working on.

pvizeli replied:
Retry logic make sense on a actor but not on a polling sensor

depl0y replied:
This is to make sure the sensors are available, even if the API isn't available when starting hass. Until hass has the right way to actually setup a platform more than once (on error), this would be easier for the user.

depl0y replied:
Also, the retry is NOT on sensor update, but on the data class that needs populating. When the data class can't populate itself, there will be no sensors to refresh.

This is because we have issue #8109

pvizeli commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,195 @@
+
+        return result
+
+    def update_device(self, device):
```
Move this code logic into entity updater

pvizeli reviewed (changes requested):
I think SCAN_INTERVAL of 3sec is to hard and consume all threads very fast. Maby default 10-30 sec and you can personal overwrite that inside platform on yaml config...

Commit:
Moved update functionality back to Entity….
Updated SCAN INTERVAL

balloob commented on tests/components/sensor/test_openhardwaremonitor.py:
```diff
@@ -0,0 +1,52 @@
+
+    def test_setup_with_config(self):
+        """Test the platform setup with configuration."""
+        self.assertTrue(
```
This test is just connecting to something that doesn't exist, and thus makes not much sense ? Also, this is already tested in the other method?

depl0y replied:
Fixed

balloob commented on tests/components/sensor/test_openhardwaremonitor.py:
```diff
@@ -0,0 +1,52 @@
+class TestOpenHardwareMonitorSetup(unittest.TestCase):
+    """Test the Open Hardware Monitor platform."""
+
+    def add_entities(self, new_entities, update_before_add=False):
```
Not sure where you copied this from, but you don't need this. Instead, just read it from the state machine.

depl0y replied:
Fixed

balloob commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,192 @@
+            self._config.get(CONF_PORT))
+
+        try:
+            response = requests.get(data_url)
```
Make sure to pass a timeout.

depl0y replied:
A 30 second timeout seems appropriate

balloob commented on tests/components/sensor/test_openhardwaremonitor.py:
```diff
@@ -0,0 +1,52 @@
+                     text=load_fixture('openhardwaremonitor.json'))
+        openhardwaremonitor.setup_platform(
+            self.hass, self.config, self.add_entities)
+        self.assertEqual(len(self.entities), 38)
```
Please check the data in the state machine: `hass.states.get('sensor.bla')`

depl0y replied:
Fixed

balloob commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,192 @@
+                child_names = names.copy()
+                child_names.append(json[OHM_NAME])
+
+                fullname = '_'.join(child_names).replace(' ', '_')
```
The name supports spaces, so please don't do this. UI looks better with spaces!

depl0y replied:
Fixed

balloob commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,192 @@
+                    obj, devices, child_path, child_names)
+
+                result = result + added_devices
+        else:
```
Instead of an else, just put a `return` at the end of the if statement. Saves 1 indentation.

depl0y replied:
Fixed, although I am more of a fan of having as little as possible exit points in a method

balloob commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,192 @@
+
+                result = result + added_devices
+        else:
+            if json[OHM_VALUE].find(' ') > -1:
```
Use a guard clause here:

```python
if json[OHM_VALUE].find(' ') == -1:
    return result

unit_of_measurement = …
```

Saves another indentation 👍

depl0y replied:
Changed, not sure this makes the code a lot more readable, but I understand the preference

balloob commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,192 @@
+
+                fullname = '_'.join(child_names).replace(' ', '_')
+
+                dev = OpenHardwareMonitorDevice(
```
You should only ever instantiate devices once. If you are fetching data again, make sure that the existing entity  picks up on the new data

depl0y replied:
Adjusted

balloob commented:
I agree that #4937 is the preferred way of doing retry logic. It will also avoid having to put some crazy hacks for mapping data to the platform.

balloob commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,192 @@
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the Open Hardware Monitor platform."""
+    data = OpenHardwareMonitorData(config, hass)
+    add_devices(data.devices, True)
```
If you have to retry, this call will just add nothing and nothing will ever be added.

depl0y replied:
Fixed by removing the retry

balloob commented on homeassistant/components/sensor/openhardwaremonitor.py:
```diff
@@ -0,0 +1,192 @@
+        """Recursively loop through child objects, finding the values."""
+        result = devices.copy()
+
+        if len(json[OHM_CHILDREN]) > 0:
```
If the length is > 0 you assume it's an array, otherwise it's a string ? 🤔

balloob replied:
Please add correct type checking.

depl0y replied:
It actually checks 2 different properties of the `json` object. First it checks if `OHM_CHILDREN` is an array, *if not* then it will check if the `OHM_VALUE` contains at least a space. The format of Open Hardware Monitor is pretty stable in this, but we don't want to be adding sensors that do not have the **`[VALUE] [UNIT_OF_VALUE]`** format.

balloob replied:
oh my bad, didn't see that they are different attributes

Commit:
Added timeout to request
Removed retry when Open Hardware Monitor API is not reachable
Fixed naming of sensors
Flow optimalisations
Fixed tests to use states

Commit:
Remove unused import

balloob commented:
Thanks! 🐬
